segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,182 @@
+
+    private final ChunkedSegmentStorage chunkedSegmentStorage;
+
+    private final ChunkedSegmentStorageConfig config;
+
+    private final AtomicBoolean closed = new AtomicBoolean();
+
+    private final AtomicBoolean suspended = new AtomicBoolean();
+
+    private final long delayInMillis;
+
+    private CompletableFuture<Void> loopFuture;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param chunkedSegmentStorage Instance of {@link ChunkedSegmentStorage}.
+     * @param config Configuration to use.
+     */
+    public GarbageCollector(ChunkedSegmentStorage chunkedSegmentStorage, ChunkedSegmentStorageConfig config) {
+        this.chunkedSegmentStorage = chunkedSegmentStorage;
+        this.config = config;
+        this.delayInMillis = Duration.ofSeconds(config.getGarbageCollectionFrequencyInSeconds()).toMillis();
+    }
+
+    /**
+     * Initializes this instance.
+     */
+    public void initialize() {
+        loopFuture = run();
+    }
+
+    public CompletableFuture<Void> run() {
+        return Futures.loop(
+                () -> !closed.get(),
+                () -> CompletableFuture.completedFuture(null)
+                        .thenComposeAsync(v -> deleteGarbage(true, config.getGarbageCollectionConcurrency()),
+                                chunkedSegmentStorage.getExecutor())
+                        .handleAsync((v, ex) -> {
+                            if (null != ex) {
+                                log.error("{} Error during run {}.", chunkedSegmentStorage.getLogPrefix(), ex);
+                            }
+                            return null;
+                        }), chunkedSegmentStorage.getExecutor());
+    }
+
+    /**
+     * Sets whether backgound claenup is suspended or not.
+     * @param value
+     */
+    void setSuspended(boolean value) {
+        suspended.set(value);
+    }
+
+    /**
+     * collect the garbage chunks.
+     *
+     * @param chunksToDelete List of chunks to delete.
+     */
+    CompletableFuture<Void> addToGarbage(Collection<String> chunksToDelete) {
+        val currentTime = System.currentTimeMillis();
+        chunksToDelete.forEach(chunkToDelete -> garbageChunks.put(chunkToDelete, currentTime));
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * Delete the garbage chunks.
+     *
+     * @param isBackground True if the caller is backgound task else False if called explicitly.
+     * @param maxItems Maximum number of items to delete at a time.
+     * @return CompletableFuture which is completed when garbage is deleted.
+     */
+    CompletableFuture<Void> deleteGarbage(boolean isBackground, int maxItems) {
+        if (suspended.get() && isBackground) {
+            return CompletableFuture.completedFuture(null);
+        }
+        val currentTime = System.currentTimeMillis();
+        val chunksToDelete = new ArrayList<String>();
+        int count = 0;
+        for (val entry : garbageChunks.entrySet()) {
```

Comment:
Why is `garbageChunks` a Map? Are you expecting to insert duplicates (that would smell like a bug).

I suggest you add a sorted list or something, sort by time, and then in this list you only need to loop through those items that need deletion (i.e, do not do an O(n) lookup every time).

Reply:
changed to priority queue